### PR TITLE
Setting up Agent manager

### PR DESCRIPTION
**Pull request by TrevorJTClarke**

### CODE REVIEW

File: manager/src/agent.rs [hunk 1]
```diff
@@ -2,12 +2,34 @@ use near_contract_standards::storage_management::StorageManagement;
...
+    Active,
+
+    // Default for any new agent, until more tasks come online
+    Pending,
+}
+
 #[derive(BorshDeserialize, BorshSerialize, Debug, Serialize, Deserialize, PartialEq)]
 #[serde(crate = "near_sdk::serde")]
 pub struct Agent {
+    pub status: AgentStatus,
```
Comment (mikedotexe) [inline manager/src/agent.rs:18]: I don't think we usually have to have `pub` before the fields here, just on the `pub struct…`

File: manager/src/agent.rs [hunk 2]
```diff
@@ -119,14 +146,46 @@ impl Contract {
             );
             let withdrawal_amount = agent_balance - storage_fee;
             agent.balance = U128::from(agent_balance - withdrawal_amount);
-            self.agents.insert(&account, &agent);
+
+            // if this is a full exit, remove agent. Otherwise, update agent
+            if remove.is_some() && remove.unwrap() {
```
Comment (mikedotexe) [inline manager/src/agent.rs:151]: I've never seen this used for a conditional, but looks fine. I almost always see the `if let Some(b)…`

File: manager/src/agent.rs [hunk 3]
```diff
@@ -119,14 +146,46 @@ impl Contract {
...
             env::panic(b"No Agent")
         }
     }
 
+    /// Removes the agent from the active & pending set of agents.
+    #[private]
+    pub fn remove_agent(&mut self, account_id: AccountId) {
+        self.agents.remove(&account_id);
+        // remove agent from agent_active_queue
+        let index = self.agent_active_queue.iter().position(|x| x == account_id);
```
Comment (mikedotexe) [inline manager/src/agent.rs:169]: neat!

File: manager/src/agent.rs [hunk 4]
```diff
@@ -101,13 +126,15 @@ impl Contract {
...
-    pub fn withdraw_task_balance(&mut self) -> Promise {
-        let account = env::predecessor_account_id();
+    /// Removes the agent from the active set of agents.
+    /// Withdraws all reward balances to the agent payable account id.
+    #[private]
+    pub fn exit_agent(&mut self, account_id: Option<AccountId>, remove: Option<bool>) -> Promise {
+        let mut account = env::predecessor_account_id();
+        if let Some(account_id) = account_id {
+            account = account_id;
+        }
```
Comment (mikedotexe) [inline manager/src/agent.rs:136]: I've seen Austin do this:
`let account = account_id.unwrap_or_else(env::predecessor_account_id);`
https://github.com/near/near-sdk-rs/blame/95954cad746502aeb64e42a8bd5bbbb1ce0df8cc/near-contract-standards/src/fungible_token/storage_impl.rs#L51

File: manager/src/owner.rs [hunk 5]
```diff
@@ -76,7 +80,7 @@ mod tests {
             .signer_account_id(accounts(3))
             .predecessor_account_id(accounts(3))
             .build());
-        contract.update_settings(None, Some(10), None, None, None, None);
+        contract.update_settings(None, Some(10), None, None, None, None, None);
```
Comment (mikedotexe) [inline manager/src/owner.rs:83]: … hey hey hey, gooodbyeeeee

File: manager/src/views.rs [hunk 6]
```diff
@@ -71,4 +84,107 @@ impl Contract {
         let task = self.tasks.get(&task_hash).expect("No task found by hash");
         task
     }
+
+    /// Gets amount of tasks alotted for a single agent per slot
+    ///
+    /// ```bash
+    /// near view cron.testnet get_total_tasks_per_agent_per_slot
+    /// ```
+    pub fn get_total_tasks_per_agent_per_slot(&self) -> u64 {
```
Comment (mikedotexe) [inline manager/src/views.rs:93]: `-> u16` might be plenty? no biggie

File: manager/src/views.rs [hunk 7]
```diff
@@ -71,4 +84,107 @@ impl Contract {
...
+        let [agent_ratio, task_ratio] = self.agent_task_ratio;
+
+        // Math example:
+        // ratio [2 agents, 5 tasks]
+        // agent can execute 5 tasks per slot
+        task_ratio.div_euclid(agent_ratio)
+    }
+}
+
+#[cfg(all(test, not(target_arch = "wasm32")))]
```
Comment (mikedotexe) [inline manager/src/views.rs:104]: as far as I know, all we need is:
`#[cfg(test)]`
but I see people using a line similar to this a lot. more of a "the more you know" thing to share

File: manager/src/agent.rs (hunk 1 above)
Comment (TrevorJTClarke) [reply manager/src/agent.rs:18]: With the file re-org, its now required for rust server

File: manager/src/agent.rs (hunk 2 above)
Comment (TrevorJTClarke) [reply manager/src/agent.rs:151]: good call

File: manager/src/views.rs (hunk 6 above)
Comment (TrevorJTClarke) [reply manager/src/views.rs:93]: done

File: manager/src/views.rs (hunk 7 above)
Comment (TrevorJTClarke) [reply manager/src/views.rs:104]: interesting

File: manager/src/utils.rs [hunk 8]
```diff
@@ -66,6 +74,67 @@ impl Contract {
...
+    }
+
+    /// Manage agents
+    fn manage_agents(&mut self) {
+        let current_slot = self.get_slot_id(None);
+        let total_agents = self.agent_active_queue.len();
+        assert!(total_agents > 0, "No agents found");
+
+        // Loop all agents to assess if really active
+        for agent_id in self.agent_active_queue.iter() {
```
[the remark this answers is not in the document]
Comment (TrevorJTClarke) [reply manager/src/utils.rs]: This made me learn some deeper things :D

File: manager/src/agent.rs (hunk 4 above)
Comment (TrevorJTClarke) [reply manager/src/agent.rs]: hahahahaahaha... ha... ah. facepalm easy

**Review by mikedotexe — APPROVED** [review]
all looks good!
`cargo fmt -- --check`
looks clean, too. Tests passing. 🚀

**Pull request merged**